_drafts/floating-point-101.md

```diff
@@ -0,0 +1,441 @@
+the _next_ immediate value (i.e. exactly 1 ULP bigger) is approx.
+[`0.3000000000000000444089`](https://www.wolframalpha.com/input?i=1.2000000000000001776357+*+2%5E%281021-1023%29). Our
+result clearly is much closer to the latter! And that's how `0.30000000000000004` is deemed the correct result.
+
+<figure>
+    <a href="https://xkcd.com/217/">
+        <img src="https://imgs.xkcd.com/comics/e_to_the_pi_minus_pi.png" title="Also, I hear the 4th root of (9^2 + 19^2/22) is pi." alt="e to the pi Minus pi" height="264" width="700" loading="lazy">
+    </a>
+    <figcaption><i>fig. 4</i> — <em>Another</em> XKCD?</figcaption>
+</figure>
+
+This category of errors, where parsing and arithmetic add up to a result a human would not expect, is why it's crucial
+NOT to use binary floating-point formats where this would be straight-up unacceptable. Prime example: finance. At the
+scale of institutions managing billions, those errors would be inevitable. The solution: **decimal floating-point
+formats**.
+
+Two such formats are in fact part of IEEE 754: `decimal64` and `decimal128`. As you can infer from those technical
+names, they have a base of 10 instead of 2, and are respectively 64 and 128 bits long. Unfortunately, their adoption
+isn't nearly as universal as that of standardized binary formats. That's largely because hardware acceleration of
+decimal floating-point arithmetic is extremely niche, so the math practically always is implemented in software –
+resulting in less pressure to use a standard _and also_ much poorer performance in comparison with binary arithmetic.
+Nevertheless, when correct handling decimal values is of utmost importance, there's no better way than decimal
+floating-point.
+
+## The There-and-Back-Again
+
+Suppose you've got a float but need to convert it to a double. To perform this transmutation, just take the existing
+exponent and significand values and pad them with zeros. Simple enough! Beware of false precision though: we've found in
+[The More-or-Less](#the-more-or-less) that a parsed value is off by a bit (but less than 0.5 ULP) _immediately_, unless
+the denominator of the original value was a power of 2. So, say, `"5.9"` parsed as a float will be printed back as
+`5.9`, as expected. Cast that to a double though and what you see is… `5.9000000953674316`. Where did the processor get
+all this extra information from? Truth is, that's what a _lack_ of information looks like. The extra digits are simply
+the initial error, made glaring because of the ULP being orders of magnitude smaller in the more precise format.
+
+The other way around – from a higher-precision format to a lower-precision one – the situation is straightforward.
+Information is unambiguously lost as the least significant digits of the significand are cut. Just one thing to watch
+out for here is the exponent being outside the target format's range – that's an example of overflow, so infinity is the
+result.
+
+## The End
+
```

Comment:
koham ciem

Reply:
okay, based on all those reacts, I really want to know what this means 😂